feat: password strength spec

iurimatias opened the pull request:
closes #15

iurimatias committed:
feat: password strength spec

iurimatias commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+  - So-So
+  - Good
+  - Great
+- **TODO**: criteria for each category
```
@corpetty @0kok0 we need to define this sort of categorization by some sort of formula and/or criteria

PascalPrecht replied:
Corey and co. prolly have much better ideas here, but one would be to go for:

```
NUM_CRITERIA = amount of criteria we have for validation
PASSWORD_CRITERIA_MATCHES = amount of matching criteria in given password

MATCHES_IN_PERCENTAGE = (PASSWORD_CRITERIA_MATCHES/NUM_CRITERIA)*100

match MATCHES_IN_PERCENTAGE:
  100 -> Great
  >= 80 -> Good
  >= 60 -> So-So
  >= 40 -> Weak
  < 40 -> Very Weak
```

OxFred replied:
> BTW, is there any restriction for using 6 characters minimum, as well as, not allowing space character? From [OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#implement-proper-password-strength-controls) the minimum length should be 8 and space char should be allowed to have a better strength password.

Yes, the consensus is recent years has been that [we should focus on length of the passwords, allow spaces (and if possible unicode) & avoid complex composition rules](https://nakedsecurity.sophos.com/2016/08/18/nists-new-password-rules-what-you-need-to-know/)

serhanwbahar replied:
Also, I'd like to add an excellent article about it:

https://www.ndss-symposium.org/wp-content/uploads/2017/09/06_3_1.pdf

serhanwbahar replied:
@noeliaSD @OxFred @John-44 

As far as I know, the best algorithm for the Password Strength Estimation is [Dropbox's zxcvbn](https://github.com/dropbox/zxcvbn).

Bitwarden is also [using zxcvbn](https://github.com/bitwarden/jslib/blob/c7ac645eb7b0cfa9c972dd382d63fab5893e1f82/src/services/passwordGeneration.service.ts#L383). So we should consider using it.

arnetheduck replied:
As does nimbus: https://github.com/status-im/nim-zxcvbn

OxFred replied:
> @noeliaSD @OxFred @John-44
> 
> As far as I know, the best algorithm for the Password Strength Estimation is [Dropbox's zxcvbn](https://github.com/dropbox/zxcvbn).
> 
> Bitwarden is also [using zxcvbn](https://github.com/bitwarden/jslib/blob/c7ac645eb7b0cfa9c972dd382d63fab5893e1f82/src/services/passwordGeneration.service.ts#L383). So we should consider using it.

Yes sure ! that's what was initially suggested on [the security-internal issue](https://github.com/status-im/security-internal/issues/56#issuecomment-984921340)

PascalPrecht commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+
+## Designs
+
+* [designs](https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=11%3A4921)
```
It would be nice to have different userflows put into screenshots here for clarity and then have a resource section where we link the source file instead.

John-44 replied:
@PascalPrecht ask and you shall be given ;-)

See userflow that starts here https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=1940%3A215207

and also the userflow that starts here https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba%E2%8E%9CDesktop?node-id=2201%3A235245

PascalPrecht commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+4.2 System checks if password is a common password (see Functional Requirements#password validation) and displays feedback (e.g "this pasword has been pwned and shouldn't be used")
+5. User types password again in the confirm password field
+6. System checks if password matches and displays "Passwords don't match" if they don't
+7. System checks if password matches and enables the "Create" button if they do
```
There's no "Create" button in this scenario

John-44 replied:
Correct, in this flow the button should be "Change Password"

PascalPrecht commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+- A password MUST BE at least 6 characters long.
+- A password MUST BE at most 64 characters long.
+- A password MUST NOT contain spaces.
+- A password MAY CONTAIN any printable ascii character.
```
Does this include Emojis?

0kok0 replied:
good question, it should and I would suggest to hint, maybe better require/suggest that from/to user. 
For the emojihash feature we require 1024 hardcoded emojis to be available for the Apps, so that drastically increases the alphabet size for the user to choose a password from, therefore allows for shorter passwords with equivalent security OR ramps up the brute force resistance with equivalent length. Also I think most rainbow tables out there don't contain emoji based password hashes yet.

John-44 replied:
the issue with supporting including emoji in a password is that we would then also have to add an additional affordance to enable users to enter emoji.  This isn't needed if we restrict ourselves to characters that can be typed with a keyboard.

PascalPrecht commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+    - If a password matches a common password, it MUST BE rejected
+    - If a password matches a common password, it MUST BE display a warning "this pasword has been pwned and shouldn't be used"
+    - If the application is online, the password MAY BE checked against an online or a local database.
+    - If the application is offline, the password MAY BE checked against a local database.
```
What is that local database? What are "common" passwords? Is this out of band? Wonder how we can be spec compliant if we don't know what database of common passwords we should refer to...

John-44 replied:
I don't think we should worry about performing this check if the application is offline, because these databases are huge.  IMHO not worth the increase in app size, or the user bandwidth to download these databases after the app is installed.  So if the application is offline, I think we can skip the 'has this password been pwned' check.

richard-ramos commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+1. User types the current password in the current password field
+2. User types the choosen password in the new password field
+3. System displays feedback on the strength of the password
+4.1 System checks if the password fits criteria (see Functional Requirements#password criteria) and displays feedback (e.g "Password is too short")
```
Should not meeting the password criteria stop you from creating an account? or is it just informative?

John-44 replied:
@richard-ramos the only mandatory criteria should be password length.  All other criteria should be just informative.

arnetheduck replied:
what drives the requirement should password length be mandatory? a 6-character password is the security equivalent of a 1-meter fence (the neat-looking wooden and nicely painted kind) and is mostly a "please-don't-look" cue.. if I'm just checking out if the app works, it's kind of annoying for no real benefit

serhanwbahar replied:
> what drives the requirement should password length be mandatory? a 6-character password is the security equivalent of a .8-meter fence (the neat-looking wooden and nicely painted kind) and is mostly a "please-don't-look" cue.. if I'm just checking out if the app works, it's kind of annoying for no real benefit

agreed. 

Good complexity with a reasonable length is the key to security. We need a balance in here for not to kill UX.

On the other hand, if the only mandatory criteria should be password length, you can see the table below what' going to happen.

![twitter_FNGoBLqXEAAn9ZI](https://user-images.githubusercontent.com/47040097/157400006-974446bc-294c-46c5-9081-133b5387f955.jpg)

richard-ramos commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+- A password MUST be checked against a list of common passwords
+    - If a password matches a common password, it MUST BE rejected
+    - If a password matches a common password, it MUST BE display a warning "this pasword has been pwned and shouldn't be used"
+    - If the application is online, the password MAY BE checked against an online or a local database.
```
Is it really necessary to check the password against an online service? It doesnt feel safe to have the password leave your device.

alaibe replied:
A local db would indeed be better, I don't think it would take much more spaces (only text) but probably something to confirm

PascalPrecht replied:
> Is it really necessary to check the password against an online service? It doesnt feel safe to have the password leave your device.

I guess the idea would be to **fetch** an online db and check against that locally. No need to send the PW over the wire

iurimatias replied:
the proposal is to use https://haveibeenpwned.com/API/v3#PwnedPasswords

offline is also ok, but can be quite limited, some of these databases are huge

John-44 replied:
> 

with the https://haveibeenpwned.com/API/v3#PwnedPasswords service, the password is never sent over the wire

OxFred replied:
Checking against previously breached passwords is indeed part of [the basic controls recommended by the NIST](https://github.com/usnistgov/800-63-3/blob/46a65dee0c99c9216ea994186748a19b23c64ae3/sp800-63b/sec5_authenticators.md?plain=1#L38)
> * Passwords obtained from previous breach corpuses.
> * Dictionary words.
> * Repetitive or sequential characters (e.g. 'aaaaaa', '1234abcd').
> * Context-specific words, such as the name of the service, the username, and derivatives thereof.

Although history has proven their recommendations should not necessarily be followed blindly, the first 3 items make a lot of sense today.

The question is should the Status app rely on a third party services for this ?
As our users machines will directly communicate with the service this should probably be an opt-in.

Also, a (weaker) alternative would be to check locally against a smaller database like https://github.com/danielmiessler/SecLists/blob/master/Passwords/Common-Credentials/10-million-password-list-top-1000000.txt

hesterbruikman commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+- it MUST be possible to view the password in the password field.
+- Everytime a password is created or changed, it MUST BE confirmed in a second password field.
+- If the password is being changed, the current password MUST BE confirmed.
+
```
Potential addition:

- Each typed character SHOULD be displayed for a split second

This is MUST for mobile, a convenience for Desktop. Depends on whether this spec will be relayed to mobile. Don't see a reason why the spec as a whole shouldn't apply to mobile.

John-44 replied:
agreed this is definitely a must for mobile.  On desktop, the user has the option to view the password while they are typing it (this option is switched off by default)

hesterbruikman commented on content/draft/general/password_strength.md:
```diff
@@ -0,0 +1,91 @@
+  - Great
+- **TODO**: criteria for each category
+
+### Misc functionality
```
Potential addition:
- Before confirming, user MUST be informed that password cannot be recovered

Create button should remain enabled (i.e. no explicit confirmation requested from user that they understand password cannot be recovered)

John-44 replied:
@hesterbruikman good idea!  Will get this added to the design :-)

noeliaSD committed:
feat: password strength spec

Replaced `Create` to `Change Password` button text in Change Password specs.
Updated minimum characters long to 10 as agreed with security team.
Specified criteria for each category: Agreed with security team to use `dropbox's zxcvbn` algorithm.
Updated some typos.

noeliaSD commented:
Please @serhanwbahar, @OxFred and @iurimatias review last commit that updates the documentation according to the discussions. The most relevant changes are:
- Updated minimum characters long to 10 as agreed with security team.
- Specified criteria for each category: Agreed with security team to use `dropbox's zxcvbn` algorithm.

If you agree, please, approve the PR and we could merge it! Thanks!!
